docs/proposals/201909_thanos_functional_sharding.md

```diff
@@ -0,0 +1,153 @@
+    
+## No Goals
+
+* Time partitioning
+* "Merging" sources together virtually for downsampling/compactions across single Source that just changes external labels.
+* Bloom filters for app metrics within blocks.
+* Allow all permutations of object storage setups:
+    * User uses multiple object storages for all sources?
+    * User uses single object storage for all sources?
+    * User uses any mix of object storages for sources. They put all in different subdirs/subpaths.
+
+This design is orthogonal for multi-buckets/multi-prefix support. 
+
+## Proposal
+
+On each component that works on the object storage (e.g Store GW and Compactor), add `--selector.relabel-config` (and corresponding `--selector.relabel-config-file`) that will
+be used to filter out what blocks should be selected for operations. Examples:
+
+* We want to run Compactor only for blocks with `external_labels` being `cluster=A`. We will run second Compactor for blocks with `cluster=B` external labels.
+
+Similar to [promtail](https://github.com/grafana/loki/blob/master/docs/promtail.md#scrape-configs) this config will follow native
+[Prometheus relabel-config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config) syntax.
+
+The relabel config will define filtering process done on **every** synchronization with object storage.
+
+We will allow manipulating with several of inputs:
+
+* `__block_id`
+* External labels:
+  * `<name>` 
+* `__block_objstore_bucket_endpoint` 
+* `__block_objstore_bucket_name`
+* `__block_objstore_bucket_path`
+
+Output:
+
+* External labels to be advertised for this block.
+* If output is empty, drop block. 
+
+By default, on empty relabel-config, all external labels are assumed.
+Intuitively blocks without any external labels will be ignored.
+
+Example usages would be:
+
+* Drop blocks which contains external labels cluster=A
+```yaml
+- action: drop
+  regex: "A"
+  source_labels:
+  - cluster
+```
+* Keep only blocks which contains external labels cluster=A
+```yaml
+- action: keep
+  regex: "A"
+  source_labels:
+  - cluster
+```
+* Drop cluster label from external labels for each blocks (if present).
+```yaml
+- action: labeldrop
+  source_labels:
+  - cluster
+```
+* Add `datacenter=ABC` external label to the result.
+```yaml
+- target_label: datacenter
+  replacement: ABC
+```
+
+Result will depends on the component. For compactor result will be ignored. For store gateway, result will traslate
+into advertise labels exposed by Store Gateway on Info method.
+
+### Work Plan
+
+* Add/import relabel config into Thanos, add relevant logic
+* Hook it for selecting blocks on Store Gateway
+    * Advertise on resulted external labels.
+* Hook it for selecting blocks on Compactor.
+    * Care must be taken with changing selection for compactor to unsure only single compactor ever running over each Source's blocks.
```

Comment:
You talked a bit about some kind of coordination between different Thanos Compact instances before in this document. Will there be any work regarding this?